File: app/javascript/channels/game_channel.js
import consumer from "./consumer"

let gameChannel;
let userId;

let containerDiv = document.getElementById('container');
if( containerDiv ) {
  userId = containerDiv.dataset['userId'];
}


let gameDiv = document.getElementsByClassName('game');
if( gameDiv.length > 0 ) {
  let gameId = gameDiv[0].dataset['gameId']
  let blackPlayerId = gameDiv[0].dataset['blackPlayerId'];
  let whitePlayerId = gameDiv[0].dataset['whitePlayerId'];

  console.log("this is user", userId, "trying watching the game ", gameId, " between ", blackPlayerId, " and ", whitePlayerId);
  gameChannel = consumer.subscriptions.create({
    channel: "GameChannel",
    game_id: gameId
  }, {
    connected() {
      console.log("connected successfully");
    },

    disconnected() {
      // Called when the subscription has been terminated by the server
    },

    received(data) {
      console.log(data);
      if( data['board'] ) {
        containerDiv.innerHTML = data['board'];
      }
      console.log("gamestate", data['game_state']);
      console.dir(`----${userId}---${blackPlayerId}----${whitePlayerId}---`);
      if(data['game_state'] == 'waiting_for_black') {
        console.log("waiting for black")
        if (userId == blackPlayerId) { 
          console.log("it is my turn, I am black");
          letterInput.style.display = 'block';
        }
      }
      if( (data['game_state'] == 'waiting_for_white') && (userId == whitePlayerId) ) {
            console.log("it is my turn, I am white");
            letterInput.style.display = 'block';
      }      
    }
  });
}

let letterInput = document.getElementById('letter');
if(letterInput) {
  const a = 97;
  const z = 123; 
  //loop through the values from i to j
  for(let l = a; l < z; l++){
    let str =String.fromCharCode(l);
    console.log(str);
    letterInput.innerHTML += `<button data-letter='${str}'>${str}</button>`;
  }
  document.querySelectorAll('#letter button').forEach(function(element){
    element.addEventListener('click', function(){
      let letter = this.dataset['letter'];
      console.log("clicked", letter);
      gameChannel.send({ guess: letter });
      letterInput.style.display = 'none';
    })
  });
}

let joinButton = document.getElementById('join');
if( joinButton ) {
  

  joinButton.addEventListener('click',function(){
    console.log("I want to join!", userId);
    gameChannel.send({ command: 'join' });
  });
}

File: app/channels/game_channel.rb
class GameChannel < ApplicationCable::Channel
  def subscribed
    stream_for game
    if current_user.present? then
      Rails.logger.warn("user #{current_user} subscribed GameChannel #{game.id} ")
      broadcast_state
    else
      Rails.logger.warn("a guest has subscribed to GameChannesl #{game.id}")
    end
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end

  def receive(data)
    if data['command'] == 'join'
      game.join(current_user)
      broadcast_state
    end
    if data['guess'].present? && current_user.id == game.current_player.user_id then
      game.guess(data['guess'])
      broadcast_state
    end
  end

  def game
    Game.find(params[:game_id])
  end

  def broadcast_state
    board = ApplicationController.render( partial: 'games/game', locals: {game: game, current_user: current_user })
    GameChannel.broadcast_to game, board: board, game_state: game.game_state
  end
end
